webui/react/src/hooks/useInputEscape.ts

```diff
@@ -0,0 +1,270 @@
+    ...document.getElementsByClassName(DRAWER_MASK_CLASSNAME),
+  ];
+
+  const overlay = overlays?.[0] as HTMLElement;
+
+  const menuBody =
+    overlay?.className === MODAL_WRAP_CLASSNAME
+      ? overlay
+      : (document.getElementsByClassName(DRAWER_BODY_CLASSNAME)?.[0] as HTMLElement);
+
+  return {
+    menuBody,
+    overlay,
+  };
+};
+
+interface InputEscape {
+  onBlur?: <T extends HTMLInputElement | HTMLTextAreaElement>(
+    e: React.FocusEvent<T> | React.KeyboardEvent<T>,
+    previousValue?: string,
+    tagID?: number,
+  ) => void;
+}
+
+interface InputTextEscape extends InputEscape {
+  inputRef?: React.RefObject<HTMLInputElement | InputRef>;
+  onFocus: <T extends HTMLInputElement | HTMLTextAreaElement>(
+    e: React.FocusEvent<T & EventTarget, Element>,
+  ) => void;
+}
+
+interface InputNumberEscape extends InputEscape {
+  onFocus: (e: React.FocusEvent<HTMLInputElement & EventTarget, Element>) => void;
+  inputRef?: Ref<HTMLInputElement>;
+}
+
+interface SelectEscape {
+  inputRef?: React.RefObject<RefSelectProps>;
+  onFocus: <T extends HTMLInputElement>(e: React.FocusEvent<T & EventTarget, Element>) => void;
+  onBlur?: (
+    e: React.FocusEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>,
+    previousValue?: string,
+    tagID?: number,
+  ) => void;
+}
+
+const onEsc = (
+  focused: boolean,
+  inputRef: React.RefObject<HTMLInputElement | AntdInputRef | RefSelectProps>,
+  event: KeyboardEvent,
+  handleFocused: (focused: boolean) => void,
+) => {
+  if (focused && event.key === 'Escape') {
+    event.stopPropagation();
+    inputRef.current?.blur();
+    getOverlayAndMenuBodyElement()?.menuBody?.focus();
+    handleFocused(false);
+  }
+};
+
+const onClick = (
+  blurred: boolean,
+  focused: boolean,
+  inputRef: React.RefObject<HTMLInputElement | AntdInputRef>,
+  event: MouseEvent,
+  handleFocused: (focused: boolean) => void,
+) => {
+  const overlayClicked =
+    (event.target as HTMLElement).className === getOverlayAndMenuBodyElement()?.overlay?.className;
+
+  if (blurred && focused && overlayClicked) {
+    event.stopPropagation();
+    handleFocused(false);
+  } else if (focused && overlayClicked) {
+    event.stopPropagation();
+    inputRef.current?.blur();
+    handleFocused(false);
+  }
+};
+
+const onClickSelect = (
+  event: MouseEvent,
+  isOpen: boolean,
+  hasOpened: boolean,
+  setHasOpened: (hasOpened: boolean) => void,
+) => {
+  const overlayClassname = getOverlayAndMenuBodyElement()?.overlay?.className;
+  if (isOpen && (event.target as HTMLElement).className === overlayClassname) {
+    event.stopPropagation();
+  }
+
+  if (hasOpened && !isOpen && (event.target as HTMLElement).className === overlayClassname) {
+    event.stopPropagation();
+    // If hasOpened is true in this instance
+    // then the event above will close the
+    // currently open Modal or Menu so we must stop the
+    // event.
+    setHasOpened(false);
+  }
```

Comment:
nit: I believe this can simplify to:
```
const isTargetOverlay = getOverlayAndMenuBodyElement()?.overlay?.className === (event.target as HTMLElement).className;
  if (isTargetOverlay && (isOpen || hasOpened)) {
    event.stopPropagation();
    setHasOpened(false);
  }
```
Double check my logic on that but I think it's equivalent.